draft-ietf-ppm-dap.md

```diff
@@ -605,46 +727,63 @@ structured as follows:
-  be the leader's, the second share should be for the first helper, and so on).
-
-To generate the report, the client begins by initializing the report nonce.
-Specifically, the client first sets `Nonce.rand` to 16 random bytes output from
-a cryptographically secure random number generator. It then sets `Nonce.time`
-to the number of seconds elapsed since the start of the UNIX epoch, rounded down to the nearest multiple of `min_batch_duration`.
-This truncation is done to ensure that the report's timestamp cannot be used to
-link the report back to the originating client.
-
-The client then finishes the report generation by sharding its measurement into
-a sequence of input shares as specified by the VDAF in use. To encrypt an input
-share, the client generates an {{HPKE}} ciphertext for the aggregator by running
+This message is called the Client's report. It consists of a "header" and the
+encrypted input share of each of the Aggregators. The header consists of:
+
+* The task ID for which the report is intended.
+
+* Information used to determine whether to include a report in a given
+  batch.
+
+    * For time-series tasks, this inclues a timestamp representing the time
+      at which the report was generated. Specifically, the `time` field is set
+      to the number of seconds elapsed since the start of the UNIX epoch. The
+      client MUST round this value to the nearest multiple of
+      `min_batch_duration` in order to ensure that that the timestamp cannot be
+      used to link a report back to the Client that generated it.
+
+* A nonce used by the Aggregators to ensure the report appears in at most one
+  batch. (See {{anti-replay}}.) The Client MUST generate this by generating 16
+  random bytes using a cryptographically secure random number gnerator.
+
+* A list of extensions to be included with the report. (See
+  {{upload-extensions}}.)
+
+To generate a report, the Client first shards its measurement into input shares
+as specified by the VDAF. It then encrypts each input share as follows:
 
 ~~~
-enc, payload = SealBase(pk, "dap-01 input share" || 0x01 || server_role,
-    task_id || nonce || extensions, input_share)
+enc, payload = SealBase(pk,
+    "dap-01 input share" || 0x01 || server_role,
+    header, input_share)
 ~~~
 
 where `pk` is the aggregator's public key; `server_role` is the Role of the
-intended recipient (`0x02` for the leader and `0x03` for the helper);
-`task_id`, `nonce`, and `extensions` are the corresponding fields of `Report`;
-and `input_share` is the aggregator's input share.
+intended recipient (`0x02` for the leader and `0x03` for the helper), `header`
+is the report header, and `input_share` is the Aggregator's input share.
+
+The order of the encrypted input shares appear MUST match the order of the
+task's `aggregator_endpoints`. That is, the first share should be the leader's,
+the second share should be for the first helper, and so on.
```

Comment:
I think L800 (a few lines down from here) needs to change. It currently reads:

> The leader MUST ignore any report whose nonce contains a timestamp that falls in
a batch interval for which it has received at least one collect request from the
collector.

Nonces no longer have a timestamp. Depending on the task configuration, reports might no longer have a timestamp at all.

More importantly, this would block an "aggressive collector" which wants to send collect requests before reports have finished arriving for a given batch, with the expectation that the collect job will stay live until the batch is "over". Supporting the aggressive collector is a nice-to-have for time-series tasks (since it breaks the need for a collector to have a clock that matches the aggregators', and removes a case where a collector can trigger data loss accidentally), and I think a requirement for fixed-chunk tasks (since batches are determined by the rate of report arrival, and the collector does not have any insight into this, so it can't know when to make another collect request).

Reply:
Good catch. I think the following solves both issues. This also matches the Helper's behavior in {{input-share-batch-validation}}.

> The Leader MUST ignore any report pertaining to a batch that has already been collected.

For time-interval tasks the intended interpretation is "the Leader rejects the report if the timestamp falls in a batch interval that has been collected". For fixed-size tasks the intended interpretation is "The Leader MUST NOT include the report in a previously collected batch."